### ISSUE TITLE
Async Downstream Correlation

### ISSUE THREAD

**Issue by jkowall** [w3c/trace-context#37]
One of the limitations of all non-commercial standards is handling of asynchronous transaction measurement and tracking. This is an advanced technique which many APM tools do not do well today. In order to do this well you have to understand the current problem:

The user makes a request to SystemA. SystemA runs several threads which spawn several new transactions. These transactions need to be tracked and attributed back to the initial user request. This poses a complex problem both on the instrumentation, downstream correlation, and reporting/visualization. 

I believe we need to add headers to allow new transactions to also be attributed to a master transaction. These could be to track threads in the process, and threads out of the process. Often times having two headers make more sense to be able to attribute them back to a specific runtime or even a second runtime running on the same (or another) system.

**Comment by yurishkuro** [comment]
You feel that trace-id and span-id do not achieve that correlation? Can you show a specific example?

**Comment by jkowall** [comment]
Sure, if you have a trace-id which goes

A->B->C

Then from C there are new threads created which go

C-> AA -> BB
C -> BB -> DD
C -> DD -> AA

You could tag all of these back to the trace-id, but we wouldn't know if they are attached to a particular thread, this means that vendor tools cannot distinguish between an async call for diagnostic purposes or what process/thread it may trace back to that eventually needs to come together to build the full picture of the transaction. This is critical for users doing diagnostics.

**Comment by yurishkuro** [comment]
It sounds like you're talking about a problem of trace context propagation across threads in a single process (C->AA, C->BB). If so, then again this is not in scope for this project whose objective is standardizing on wire formats. The instrumentation concerns are the raison d'être of OpenTracing API, which (for Java) has a notion of "active span" that can be carried over thread boundaries.

**Comment by jcchavezs** [comment]
Can `baggage` be what you are looking for? https://github.com/TraceContext/tracecontext-spec/issues/13

**Comment by codefromthecrypt** [comment]
Hi, jonah.

To try and tease this out.

I know in the past we had folks in APM who have a notion of "next async id"
ex naver pinpoint do, which help to establish async scheduling order
independent of execution time. They use this so that profiler style
functions can operate. I am not sure one of the use cases of detecting a
loop (as you are noticing here) are supported. Nod if this is relevant..

Moreover the actual thread vs the logical one is an interesting and tricky
thing especially in apps such as reactive who switch many times. In such
case you need to or might want to know the entrypoint into that reactive
flow (flux). I know some save call stack references and dedupe by hash of
them.. not sure if this hash is propagated (so as to detect a loop). Nod if
relevant...

One last this is frankly most open source tracers do nothing very fancy wrt
propagation.. a lot of time that code is closed source or buried in a
larger code base that is more APM in nature. Personally, I would love to
see more advanced propagation apis, but many have never seen them. If you
have design ideas please mention.. it is indeed important to know the use
and design cases behind context decisions. It is also important imho to
have more relevant tools in OSS if we ever would like vendors to actually
use them ;)

Regardless. THANK YOU!

**Comment by jkowall** [comment]
@jcchavezs Once again baggage will become a pile of junk versus building critical tracing technologies into the design. These are key things we should define in the spec. 

@adriancole Thanks for the feedback and reaching out! I'm not sure if detecting a loop is the main use case here, we would actually detect it as stalled and show where the stall is occurring. 

I do believe the analysis of threads and hence being able to do distributed thread contention and timing is the main reason for this type of header. We had to add this into our design later in life when we added async detection and support. We actually show async differently and provide detailed analytics on these transactions. It's key in microservices architecture when you are trying to trace specific sub-requests back to a user/API transaction. 

Right now OSS is early days, but give it a couple years and you'll need this :)
